cmd/juju/application/utils/origin.go

```diff
@@ -17,11 +17,7 @@ import (
 // DeduceOrigin attempts to deduce the origin from a channel and a platform.
 // Depending on what the charm URL schema is, will then construct the correct
 // origin for that application.
-func DeduceOrigin(url *charm.URL, channel charm.Channel, platform corecharm.Platform) (commoncharm.Origin, error) {
-	if url == nil {
-		return commoncharm.Origin{}, errors.NotValidf("charm url")
-	}
-
+func DeduceOrigin(schema string, revision int, channel charm.Channel, platform corecharm.Platform) (commoncharm.Origin, error) {
```

Comment:
```suggestion
func MakeOrigin(schema charm.Schema, revision int, channel charm.Channel, platform corecharm.Platform) (commoncharm.Origin, error) {
```

I have concerns about this function. It's moved from its original design, which I'm fine with, but it's no longer deducing an origin, instead it's a origin factory that's created based on a set of inputs. With that in mind the schema should be typed as it's the typed version everywhere but refresh. Moving refresh to the type with strengthen this method.